chpl/chpl-service/src/main/java/gov/healthit/chpl/listener/MissingReasonListener.java

```diff
@@ -0,0 +1,100 @@
+import org.aspectj.lang.annotation.Before;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
+
+import gov.healthit.chpl.dao.CertifiedProductDAO;
+import gov.healthit.chpl.domain.CertifiedProductSearchDetails;
+import gov.healthit.chpl.domain.ListingUpdateRequest;
+import gov.healthit.chpl.dto.questionableActivity.QuestionableActivityListingDTO;
+import gov.healthit.chpl.entity.CertificationStatusType;
+import gov.healthit.chpl.exception.EntityRetrievalException;
+import gov.healthit.chpl.exception.MissingReasonException;
+import gov.healthit.chpl.manager.CertifiedProductDetailsManager;
+import gov.healthit.chpl.questionableactivity.ListingQuestionableActivityProvider;
+import gov.healthit.chpl.util.ErrorMessageUtil;
+
+@Component
+@Aspect
+public class MissingReasonListener {
+
+    private ErrorMessageUtil errorMessageUtil;
+    private CertifiedProductDetailsManager cpdManager;
+    private ListingQuestionableActivityProvider listingQuestionableActivityProvider;
+
+    /**
+     * Autowired constructor for dependency injection.
+     * 
+     * @param errorMessageUtil
+     *            - Error message utility class
+     * @param cpdManager
+     *            - CertifiedProductDetailsManager
+     * @param listingDao
+     *            - CertifiedProductDAO
+     * @param listingQuestionableActivityProvider
+     *            - ListingQuestionableActivityProvider
+     */
+    @Autowired
+    public MissingReasonListener(final ErrorMessageUtil errorMessageUtil,
+            final CertifiedProductDetailsManager cpdManager, final CertifiedProductDAO listingDao,
+            final ListingQuestionableActivityProvider listingQuestionableActivityProvider) {
+
+        this.errorMessageUtil = errorMessageUtil;
+        this.cpdManager = cpdManager;
+        this.listingQuestionableActivityProvider = listingQuestionableActivityProvider;
+    }
+
+    /**
+     * Looks for reason for listing update if required.
+     * 
+     * @param updateRequest
+     *            the listing update object
+     * @throws EntityRetrievalException
+     *             if the listing cannot be found
+     * @throws MissingReasonException
+     *             if a reason was required but is not found
+     */
+    @Before("execution(* gov.healthit.chpl.manager.CertifiedProductManager+.update(..)) && "
+            + "args(.., updateRequest))")
+    public void checkReasonProvidedIfRequiredOnListingUpdate(final ListingUpdateRequest updateRequest)
+            throws EntityRetrievalException, MissingReasonException {
+        CertifiedProductSearchDetails newListing = updateRequest.getListing();
+        CertifiedProductSearchDetails origListing = cpdManager.getCertifiedProductDetails(newListing.getId());
+        List<QuestionableActivityListingDTO> activities;
+
+        QuestionableActivityListingDTO activity = listingQuestionableActivityProvider
```

Comment:
Not sure why this class is showing as new since it (or something like it) already existed but I recently had a ticket that I think has been resolved and merged to staging that added a check behind a flag that made all 2014 listing updates require a reason (same as 2011 listing updates currently). Please make sure that new code is integrated here.

Reply:
The original `MissingReasonListener` existed in the API project.  With the refactoring of the code from the controller to the manager, I had to move part of the functionality in `MissingReasonListener' to the service project.  So, there are 2 'MissingReasonListener' classes - one in API and one in Service.

Reply:
That makes sense but I guess why didn't the one from API just get moved into Service and have the pointcut changed? Now there's basically duplicate code (except the flagged 2014 check) right?

Reply:
And regardless if you are supposed to be able to call the manager method and expect the exact same behavior as calling the controller endpoint, I think the new listener in the service project should be identical to the one in API. Which goes back to just moving the whole thing instead of having 2.